### PR TITLE
Data-based validation

### PR DESCRIPTION
**Pull request by hayfield**
This comes at validation in a manner that explicitly looks at the values in a Dataset.

### CODE REVIEW

**Comment by hayfield** [comment]
Need to deal with `complete="0"` codelists

File: .gitignore [hunk 1]
```diff
@@ -6,8 +6,13 @@
 pyenv/
 pyenv2/
 
+# tool-generated folders
+prof/
+IATI-Data-Snapshot/
```
Comment (dalepotter) [inline .gitignore]: Not sure we should be adding a tool folder to the `.gitignore`. This sounds more like a usage of the library, rather than part of the library itself.

File: .gitignore [hunk 2]
```diff
@@ -6,8 +6,13 @@
...
+
 # defined files to ignore
 ignore_*
+test.py
```
Comment (dalepotter) [inline .gitignore]: Not sure what this is for?

File: iati/core/default.py [hunk 3]
```diff
@@ -8,7 +8,10 @@
     Implement more than Codelists.
 
 """
+
+from collections import defaultdict
```
Comment (dalepotter) [inline iati/core/default.py]: Imports with a `from` keyword should be placed after those with a straight `import` statement:
- https://google.github.io/styleguide/pyguide.html?showone=Imports_formatting#Imports_formatting
- https://stackoverflow.com/a/20763446/2761030

File: iati/core/default.py [hunk 4]
```diff
@@ -102,6 +105,73 @@ def codelists(version=None, use_cache=False):
...
+        iati.core.ruleset.Ruleset: The default Ruleset for the specified version of the Standard.
+
+    Raises:
+        ValueError: When a specified version is not a valid version of the IATI Standard.
+
+    Todo:
+        Actually handle versions, including errors.
+
+    """
+    name = 'standard_ruleset'
```
Comment (dalepotter) [inline iati/core/default.py]: This should be defined as a constant in `resources.py`

File: iati/core/default.py [hunk 5]
```diff
@@ -102,6 +105,73 @@ def codelists(version=None, use_cache=False):
...
+
+    path = iati.core.resources.get_ruleset_path(name, version)
+    ruleset_str = iati.core.resources.load_as_string(path)
+
+    return iati.core.rulesets.Ruleset(ruleset_str)
+
+def ruleset_schema(version=None):
+    name = 'ruleset_schema'
```
Comment (dalepotter) [inline iati/core/default.py]: As with the `standard_ruleset`, this should be defined as a constant in `resources.py`.

File: iati/core/resources.py [hunk 6]
```diff
@@ -191,6 +201,42 @@ def get_codelist_path(codelist_name, version=None):
...
+def get_codelist_mapping_path(version=None):
+    """Determine the path of the Codelist mapping file.
+
+    version (str): The version of the Standard to return the data files for. Defaults to None. This means that the path is returned for a filename at the latest version of the Standard.
+
+    Returns:
+        str: The path to a file containing the mapping file.
+
+    Todo:
+        Test this.
```
Comment (dalepotter) [inline iati/core/resources.py]: Can be tested with the `*standard_version_optional` fixture: https://github.com/IATI/iati.core/blob/dev/iati/core/tests/utilities.py#L131-L144

File: iati/core/resources.py [hunk 7]
```diff
@@ -236,9 +282,72 @@ def get_test_data_path(name, version=None):
...
+        str: The path to a file containing the specified test Ruleset.
+
+    Note:
+        Does not check whether the specified file actually exists.
+
+    Warning:
+        Needs to handle a more complex file structure than a single flat directory.
+
+    Todo:
+        Test this.
```
Comment (dalepotter) [inline iati/core/resources.py]: Can be tested with the *standard_version_optional fixture: https://github.com/IATI/iati.core/blob/dev/iati/core/tests/utilities.py#L131-L144

File: iati/core/resources.py [hunk 8]
```diff
@@ -261,6 +370,50 @@ def get_folder_name_for_version(version=None):
...
+        str: The path to a file containing the specified schema.
+
+    Note:
+        Does not check whether the specified schema actually exists.
+
+    Warning:
+        Further exploration needs to be undertaken in how to handle multiple versions of the Standard.
+
+    Todo:
+        Handle versions of the standard other than 2.02.
```
Comment (dalepotter) [inline iati/core/resources.py]: When merged into dev, this should work, since the underlaying `iati.core.resources.get_folder_name_for_version` is tested against versions `1.04` -> `2.02`.

File: iati/core/tests/test_default.py [hunk 9]
```diff
@@ -48,6 +48,40 @@ def test_default_codelists(self):
...
+
+    def test_codelist_mapping_xpath(self):
+        """Check that the Codelist mapping file is being read correctly.
+
+        Todo:
+            Split into multiple tests.
+        """
+        mapping = iati.core.default.codelist_mapping()
+
+        assert mapping['Version'][0]['xpath'] == '//iati-activities/@version'
```
Comment (dalepotter) [inline iati/core/tests/test_default.py]: Would also be good to test for codelists which have more than one mapping. In the case of the `Version` codelist, this maps to `//iati-activities/@version` and `//iati-organisations/@version`. The `Sector` codelist is another with multiple mappings.

File: iati/core/tests/test_resources.py [hunk 10]
```diff
@@ -113,6 +113,12 @@ def test_find_schema_paths_file_extension(self, get_schema_path_function):
         for path in paths:
             assert path[-4:] == iati.core.resources.FILE_SCHEMA_EXTENSION
 
+    def test_get_test_data_paths_in_folder(self):
+        """Check that test data is being found in specified subfolders."""
+        paths = iati.core.resources.get_test_data_paths_in_folder('ssot-activity-xml-fail')
+
+        assert len(paths) == 237
```
Comment (dalepotter) [inline iati/core/tests/test_resources.py]: Wondering if some of these 'number of expected files' should be consolidated somewhere, rather than defined as magic numbers - perhaps in `iati/core/tests/utilities.py`, or elsewhere, even a new file: `iati/core/tests/constants.py` ??

File: iati/core/default.py (hunk 3 above)
Comment (hayfield) [reply iati/core/default.py]: More specific reference: https://github.com/reddit/reddit/wiki/PythonImportGuidelines

File: iati/core/default.py (hunk 4 above)
Comment (hayfield) [reply iati/core/default.py]: This is something to fix in a branch forked from `ruleset-implementation`

File: iati/core/default.py (hunk 5 above)
Comment (hayfield) [reply iati/core/default.py]: Noted in https://github.com/IATI/iati.core/pull/49#issuecomment-327143653

File: iati/core/resources.py (hunk 6 above)
Comment (hayfield) [reply iati/core/resources.py]: These various things in `resources.py` are present in the `dev` branch - suggest that they should be fixed in a branch forked from there.

Comment (dalepotter) [reply iati/core/resources.py]: Sounds sensible.

File: iati/core/tests/test_default.py (hunk 9 above)
Comment (hayfield) [reply iati/core/tests/test_default.py]: Relates to the validation tests for the `valid_iati_invalid_codes_multiple_xpaths_for_codelist_[first|second].xml` files

Comment (hayfield) [reply iati/core/tests/test_default.py]: https://github.com/IATI/iati.core/pull/142/commits/268a9b5292364ac88c6ae6547344a97799803aa7

File: .gitignore (hunk 1 above)
Comment (hayfield) [reply .gitignore]: https://github.com/IATI/iati.core/pull/142/commits/0881dc3c0ae70477752e7be527c37d7539332907

File: .gitignore (hunk 2 above)
Comment (hayfield) [reply .gitignore]: https://github.com/IATI/iati.core/pull/142/commits/0881dc3c0ae70477752e7be527c37d7539332907

File: iati/core/default.py (hunk 3 above)
Comment (hayfield) [reply iati/core/default.py]: https://github.com/IATI/iati.core/pull/142/commits/3cf58da9c1d903eb3ddc4cc46a679ab4a3fef0dd

File: iati/core/tests/test_resources.py (hunk 10 above)
Comment (hayfield) [reply iati/core/tests/test_resources.py]: Once multi-version support is added, probably https://github.com/IATI/iati.core/pull/142/commits/7bc397fd2303f6b87a9c486178a8ede36891b7f8

File: iati/core/default.py (hunk 5 above)
Comment (allthatilk) [reply iati/core/default.py]: done

File: .gitignore [hunk 11]
```diff
@@ -6,6 +6,7 @@
 pyenv/
 pyenv2/
 
+
```
Comment (hayfield) [inline .gitignore]: Remove line

File: iati/tests/test_data.py [hunk 12]
```diff
@@ -30,6 +30,11 @@ def test_dataset_no_params(self):
 
         assert ('__init__() missing 1 required positional argument' in str(excinfo.value)) or ('__init__() takes exactly 2 arguments' in str(excinfo.value))
 
+    def test_dataset_empty_string(self):
+        """Test Dataset creation with a valid XML string that is not IATI data."""
```
Comment (hayfield) [inline iati/tests/test_data.py]: Incorrect docsting

File: CHANGELOG.md [hunk 13]
```diff
@@ -8,14 +8,54 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) a
...
+
+- [Resources] Add method to load data files relating to pyIATI (rather than the IATI Standard, or tests). [#45]
+- [Resources] Allow test files to be located within sub-folders by including slashes (`/`) in the name. [#45]
+- [Resources] Detect encoding of files that are not UTF-8. [#45]
+
+- [Validation] Change from `validate.py` to `validator.py` to improve readability of code using this module. [#86]
+- [Validation] Add a YAML file containing error information. [#117]
+- [Validation] Check whether a string is valid XML - truthy. [#45]
+- [Validation] Check whether a string is valid XML - detailed error information. [#45]
+- [Validation] Provide custom error messages for lxml errors stating that a string is not XML caused by published IATI data. [#90]
```
Comment (allthatilk) [inline CHANGELOG.md]: Can this be rephrased? It's hard to understand atm. Something like: `Provide custom error messages when lxml returns errors for data that is not XML in published IATI data.`

File: CHANGELOG.md [hunk 14]
```diff
@@ -8,14 +8,54 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) a
...
+- [Validation] Check whether a Dataset is valid against an IATI Schema - truthy. [#45]
+- [Validation] Check whether a Dataset is valid against an IATI Schema - detailed error information. [#45]
+- [Validation] Provide custom error messages for lxml errors stating that a Dataset is not IATI XML caused by published IATI data. [#92]
```
Comment (allthatilk) [inline CHANGELOG.md]: Can this be rephrased? It's hard to understand atm. Something like: `Provide custom error messages when lxml returns errors for data that is not IATI XML in published IATI data.`

File: iati/codelists.py [hunk 15]
```diff
@@ -209,8 +215,13 @@ def __eq__(self, other):
         Todo:
             Utilise all attributes as part of the equality process.
 
+            Test comparison with strings.
+
         """
-        return ((self.name) == (other.name)) and ((self.value) == (other.value))
+        if isinstance(other, iati.codelists.Code):
```
Comment (allthatilk) [inline iati/codelists.py]: Consider making this block more pythonic, but optional whether this is worth changing now.

File: iati/resources.py [hunk 16]
```diff
@@ -184,12 +189,48 @@ def get_codelist_path(codelist_name, version=None):
...
+def get_codelist_mapping_path(version=None):
+    """Determine the path of the Codelist mapping file.
+
+    version (str): The version of the Standard to return the data files for. Defaults to None. This means that the path is returned for a filename at the latest version of the Standard.
+
+    Returns:
+        str: The path to a file containing the mapping file.
+
+    Todo:
+        Test this.
```
Comment (allthatilk) [inline iati/resources.py]: Why isn't this tested?

File: iati/resources.py [hunk 17]
```diff
@@ -184,12 +189,48 @@ def get_codelist_path(codelist_name, version=None):
...
+        name (str): The name of the data file to locate. The name must include the file extension.
+
+    Returns:
+        str: The path to the specified file.
+
+    Note:
+        Does not check whether the specified file actually exists.
+
+    Todo:
+        Test this.
```
Comment (allthatilk) [inline iati/resources.py]: Why isn't this tested?

File: iati/resources.py [hunk 18]
```diff
@@ -396,8 +475,20 @@ def load_as_string(path):
     Todo:
         Pass in PACKAGE as a default parameter, so that this code can be used by other library modules (e.g. iati.fetch).
 
+        Add test to load a dataset saved in non-UTF-8 formats. This should include `UTF-16LE`, `UTF-16BE` and `windows-1252` since there are published Datasets using these encodings.
```
Comment (allthatilk) [inline iati/resources.py]: Why hasn't this test been done?

File: iati/resources/lib_data/README.md [hunk 19]
```diff
@@ -0,0 +1,4 @@
+Files containing data required by pyIATI.
+
+* The files are not part of the IATI Standard itself.
```
Comment (allthatilk) [inline iati/resources/lib_data/README.md]: remove `itself`

File: iati/resources/lib_data/validation_err_codes.yaml [hunk 20]
```diff
@@ -0,0 +1,358 @@
+# A YAML file containing the error codes that may be output from validation.
+#
+# The structure of a code is:
+#
+#   - name-of-the-error-or-warning:
```
Comment (dalepotter) [inline iati/resources/lib_data/validation_err_codes.yaml]: Could add back ticks to the key names. This should signify that they are the keys and what follows is a definition.

File: CHANGELOG.md (hunk 13 above)
Comment (hayfield) [reply CHANGELOG.md]: https://github.com/IATI/pyIATI/pull/209/commits/6b61368da70b82f1ffc74ecb2987c93f481c4447

File: CHANGELOG.md (hunk 14 above)
Comment (hayfield) [reply CHANGELOG.md]: https://github.com/IATI/pyIATI/pull/209/commits/6b61368da70b82f1ffc74ecb2987c93f481c4447

File: iati/resources/lib_data/README.md (hunk 19 above)
Comment (hayfield) [reply iati/resources/lib_data/README.md]: https://github.com/IATI/pyIATI/pull/209/commits/ccc90884f4c8b934088335b0491e373ff0783723

File: iati/codelists.py (hunk 15 above)
Comment (hayfield) [reply iati/codelists.py]: https://github.com/IATI/pyIATI/pull/209/commits/3050de8639068bb154d17c77d997d5342ad18e85

File: iati/tests/test_data.py [hunk 21]
```diff
@@ -218,33 +236,48 @@ def test_instantiation_dataset_from_string_with_encoding(self, encoding):
...
         ("UTF-16", "BIG5"),
-        ("UTF-16", "EUC-JP"),
-        ("UTF-32", "UTF-16"),
-        ("ASCII", "UTF-16"),
-        ("ISO-8859-1", "UTF-16"),
-        ("ISO-8859-2", "UTF-16"),
-        ("BIG5", "UTF-16"),
-        ("EUC-JP", "UTF-16")])
-    def test_instantiation_dataset_from_string_with_encoding_mismatch(self, encoding_declared, encoding_used):
+        ("UTF-16", "EUC-JP")
```
Comment (dalepotter) [inline iati/tests/test_data.py]: Not sure why `UTF-32` (amongst others) has been removed here???

File: iati/resources.py (hunk 16 above)
Comment (hayfield) [reply iati/resources.py]: https://github.com/IATI/pyIATI/commit/fca1fb0864057049606b6798ab351177e57e038f

File: iati/tests/test_validator.py [hunk 22]
```diff
@@ -0,0 +1,960 @@
...
+        assert len(error_log_mixed_contents) == 2
+        assert error_log_mixed_contents.contains_errors()
+        assert error_log_mixed_contents.contains_warnings()
+        assert error_log_mixed_contents.contains_error_called(err_name)
+        assert error_log_mixed_contents.contains_error_called(warning_name)
+        assert error_log_mixed_contents.contains_error_of_type(err_type)
+        assert error_log_mixed_contents.contains_error_of_type(warning_type)
+        assert not error_log_mixed_contents.contains_error_of_type(unused_exception_type)
+
+    @pytest.mark.parametrize("not_validatio_error_obj", iati.tests.utilities.generate_test_types([], True))
```
Comment (dalepotter) [inline iati/tests/test_validator.py]: Typo `not_validatio_error_obj`

File: iati/validator.py [hunk 23]
```diff
@@ -0,0 +1,730 @@
...
+        # format error messages with context-specific info
+        try:
+            self.help = self.help.format(**calling_locals)
+            self.info = self.info.format(**calling_locals)
+        except KeyError:  # as missing_var_err:
+            # raise NameError('The calling scope must contain a `{missing_var_err.args[0]}` variable for providing information for the error message.'.format(**locals()))
+            pass
+
+        # set general attributes for this type of error that require context from the calling scope
+        # TODO: Determine what the defaults should be should the appropriate values not be available  # pylint: disable=fixme
```
Comment (dalepotter) [inline iati/validator.py]: Is this still relevant for right now?

File: iati/validator.py [hunk 24]
```diff
@@ -0,0 +1,730 @@
...
+    Raises:
+        KeyError: When a specified base_exception is not a valid type of exception.
+
+    Todo:
+        Raise the correct error for incorrect base_exception values.
+        Raise an error when there is a problem with non-base_exception-related errors.
+
+    """
+    err_codes_str = iati.resources.load_as_string(iati.resources.get_lib_data_path('validation_err_codes.yaml'))
+    err_codes_list_of_dict = yaml.safe_load(err_codes_str)
```
Comment (dalepotter) [inline iati/validator.py]: is it not easier to just to:

`yaml.safe_load(iati.resources.get_lib_data_path('validation_err_codes.yaml'))` ?

File: iati/resources.py (hunk 18 above)
Comment (hayfield) [reply iati/resources.py]: 57536c9793d6891978be81eca682ae3312b0da8c

File: iati/tests/test_validator.py (hunk 22 above)
Comment (hayfield) [reply iati/tests/test_validator.py]: https://github.com/IATI/pyIATI/pull/209/commits/e603617171e8b0955003d15acbb6b15031d3ff66

File: iati/tests/test_data.py (hunk 21 above)
Comment (hayfield) [reply iati/tests/test_data.py]: See: The `Note:` in the docstring:

```
There are a number of other errors that may be raised with alternative encoding mismatches. These are not supported since it does not appear likely enough that they will occur and be a large issue in practice.

This is due to a pair of issues with libxml2 (the underlying library behind lxml):

1. It only supports a limited number of encodings out-of-the-box.
2. Different encoding pairs (whether supported or unsupported by libxml2; byte-equivalent-subsets or distinct encodings; and more), will return different error codes in what one would expect to act as equivalent situations.
```

File: iati/resources/lib_data/validation_err_codes.yaml (hunk 20 above)
Comment (hayfield) [reply iati/resources/lib_data/validation_err_codes.yaml]: IMO, that'd confuse things because it'd look like you're supposed to include the backticks.

If doing anything along these lines, I'd look to make it clear what the descriptions are (since these are the things to be culled). I'm not, however, sure if there's a nice way to do this.

File: iati/validator.py (hunk 23 above)
Comment (hayfield) [reply iati/validator.py]: It's still a TODO, though requires more work than there is time for on this PR.

https://github.com/IATI/pyIATI/pull/209/commits/f7e4246398256558a379adc2691763b4ae690614 moves this and a couple of other inline TODOs to docstrings.

File: iati/validator.py (hunk 24 above)
Comment (hayfield) [reply iati/validator.py]:
```
>>> yaml.safe_load(iati.resources.get_lib_data_path('validation_err_codes.yaml'))
'resources/lib_data/validation_err_codes.yaml'
```

Seemingly not

**Comment by hayfield** [comment]
@allthatilk @dalepotter Requested changes made in #209 and #210

File: CHANGELOG.md [hunk 25]
```diff
@@ -22,10 +22,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) a
 - [Validation] Add a YAML file containing error information. [#117]
 - [Validation] Check whether a string is valid XML - truthy. [#45]
 - [Validation] Check whether a string is valid XML - detailed error information. [#45]
-- [Validation] Provide custom error messages for lxml errors stating that a string is not XML caused by published IATI data. [#90]
+- [Validation] Provide custom error messages when lxml returns errors for a string that is not XML. [#90]
 - [Validation] Check whether a Dataset is valid against an IATI Schema - truthy. [#45]
 - [Validation] Check whether a Dataset is valid against an IATI Schema - detailed error information. [#45]
-- [Validation] Provide custom error messages for lxml errors stating that a Dataset is not IATI XML caused by published IATI data. [#92]
+- [Validation] Provide custom error messages when lxml returns errors for Datasets that is do not contain valid IATI XML. [#92]
```
Comment (allthatilk) [inline CHANGELOG.md]: "that is do not" -> `that do not`

File: iati/resources/lib_data/validation_err_codes.yaml (hunk 20 above)
Comment (dalepotter) [reply iati/resources/lib_data/validation_err_codes.yaml]: Hrmm, not sure either way is perfect so let's not make any further changes at the moment.

File: CHANGELOG.md (hunk 25 above)
Comment (hayfield) [reply CHANGELOG.md]: 72ee53f0c7be8b9e45a3bfead1cf8106787dc87b

File: iati/tests/test_data.py (hunk 21 above)
Comment (dalepotter) [reply iati/tests/test_data.py]: I see!

**Comment by andylolz** [comment]
@hayfield I have a question about 6d39e031 but not sure the best place to ask!

Is it necessary to specify the xml namespace somewhere in order for `xpath.attrib['xml:lang']` to work?

https://github.com/IATI/pyIATI/blob/6d39e031c3a6196ffa0516aa4614300338adc9a7/iati/validator.py#L303

I’m not sure why, but it doesn’t work for me :( I’m definitely using lxml v4.1.0, though…

**Comment by hayfield** [comment]
@andylolz IIRC, that fix worked to prevent an `XPathEvalError` - all other XPaths in the mapping file start with `iati-activity/` or `iati-organisation/`, while this does not. The super-fun-string-manip deals with the edge case.

https://github.com/IATI/pyIATI/blob/f3c9b54dd3fb6d5e6fc48b3329e6b9a9682d1318/iati/tests/test_validator.py#L707-L713 indicates that there is a working test case for the `Language` Codelist, though the [test file](https://github.com/IATI/pyIATI/blob/f3c9b54dd3fb6d5e6fc48b3329e6b9a9682d1318/iati/resources/test_data/202/valid_iati.xml) doesn't actually contain the attribute. Also, the SSOT-related tests only run truthy Schema (rather than content) checks. This would indicate that it might not actually work as intended since the tests don't appear as comprehensive as they should be. Will add a note to investigate further in the morning.

**Comment by andylolz** [comment]
Aha! That’s great – thanks for investigating, @hayfield!

**Comment by hayfield** [comment]
@andylolz #226 adds a fix for this, and #227 creates an Issue to log potentially related problems with namespaces
